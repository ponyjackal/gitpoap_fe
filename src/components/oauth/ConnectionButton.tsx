import React from 'react';
import styled from 'styled-components';
import { useClaimContext } from '../claims/ClaimContext';
import { rem } from 'polished';
import { Button, ClaimCircle } from '../shared/elements';
import { useRouter } from 'next/router';
import { useUser } from '../../hooks/useUser';
import { GitPOAP } from '../shared/elements/icons';
import { useGetEmail } from '../../hooks/useGetEmail';

const Content = styled.div`
  display: flex;
  flex-direction: row;
  justify-content: center;
  align-items: center;
  color: white;
`;

const GitPOAPIcon = styled(GitPOAP)`
  path {
    fill: white;
  }
`;

type Props = {
  className?: string;
  hideText?: boolean;
};

export const ConnectionButton = ({ className, hideText }: Props) => {
  const { claimedIds, userClaims, setIsOpen } = useClaimContext();
  const user = useUser();
  const userClaimCount = userClaims?.length;
  const router = useRouter();
  const email = useGetEmail({ showNotification: false });

  if (user === null) {
    return (
      <Content className={className}>
        <Button onClick={() => router.push('/eligibility')} leftIcon={!hideText && <GitPOAPIcon />}>
          {hideText ? <GitPOAPIcon /> : 'Check Eligibility'}
        </Button>
      </Content>
    );
  }

  /* If user has any claims at all, then render the following regardless of connections */
  if (userClaimCount && userClaimCount > 0 && userClaimCount - claimedIds.length > 0) {
    /* Connected to GitHub, but HAS open claims */
    const netClaims = userClaimCount - claimedIds.length;
    return (
      <Content className={className}>
        <Button
          onClick={() => setIsOpen(true)}
          leftIcon={!hideText && <GitPOAPIcon />}
          rightIcon={
            !hideText && <ClaimCircle key={`claim-circle-${netClaims}`} value={netClaims} />
          }
        >
          {hideText ? (
            <ClaimCircle key={`claim-circle-${netClaims}`} value={netClaims} />
          ) : (
            'VIEW & MINT'
          )}
        </Button>
      </Content>
    );
  }

  /* User has no connected accounts */
  if (!user?.capabilities.hasGithub && !email) {
    return (
      <Content className={className}>
        <Button
          onClick={() => router.push(`/settings#integrations`)}
          leftIcon={!hideText && <GitPOAPIcon />}
        >
          {hideText ? <GitPOAPIcon /> : 'Connect Accounts'}
        </Button>
      </Content>
    );
  }

  /* Has at least ONE connection, but NO open claims */
  return (
    <Content className={className}>
      <Button
        onClick={() => setIsOpen(true)}
        variant="outline"
        leftIcon={!hideText && <GitPOAPIcon />}
        sx={{ minWidth: hideText ? 0 : rem(125) }}
      >
        {hideText ? <GitPOAPIcon /> : 'NONE TO MINT'}
      </Button>
    </Content>
  );
};
