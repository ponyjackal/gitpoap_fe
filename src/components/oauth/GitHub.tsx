import React from 'react';
import styled from 'styled-components';
import { useClaimContext } from '../claims/ClaimContext';
import { rem } from 'polished';
import { GoMarkGithub } from 'react-icons/go';
import { Button, ClaimCircle } from '../shared/elements';
import { useRouter } from 'next/router';
import { useFeatures } from '../FeaturesContext';
import { useUser } from '../../hooks/useUser';
import { GitPOAP } from '../shared/elements/icons';

const Content = styled.div`
  display: flex;
  flex-direction: row;
  justify-content: center;
  align-items: center;
  color: white;
`;

const ConnectedButton = styled(Button)<{ hideText?: boolean }>`
  min-width: ${(props) => (props.hideText ? 0 : rem(125))};
`;

const GitPOAPIcon = styled(GitPOAP)`
  path {
    fill: white;
  }
`;

type Props = {
  className?: string;
  hideText?: boolean;
};

export const GitHub = ({ className, hideText }: Props) => {
  const { claimedIds, userClaims, setIsOpen } = useClaimContext();
  const user = useUser();
  const { hasCheckEligibility } = useFeatures();
  const userClaimCount = userClaims?.length;
  const router = useRouter();

  /* User has no connected GitHub account */
  if (!user?.capabilities.hasGithub) {
    return user === null && hasCheckEligibility ? (
      <Content className={className}>
        <Button onClick={() => router.push('/eligibility')} leftIcon={!hideText && <GitPOAPIcon />}>
          {hideText ? <GitPOAPIcon /> : 'Check Eligibility'}
        </Button>
      </Content>
    ) : (
      <Content className={className}>
        <Button
          onClick={() => router.push(`/settings#integrations`)}
          leftIcon={!hideText && <GoMarkGithub size={16} />}
        >
          {hideText ? <GoMarkGithub size={16} /> : 'Connect GitHub'}
        </Button>
      </Content>
    );
  }

  if (userClaimCount && userClaimCount > 0 && userClaimCount - claimedIds.length > 0) {
    /* Connected to GitHub, but HAS open claims */
    const netClaims = userClaimCount - claimedIds.length;
    return (
      <Content className={className}>
        <Button
          onClick={() => setIsOpen(true)}
          leftIcon={!hideText && <GoMarkGithub size={16} />}
          rightIcon={
            !hideText && <ClaimCircle key={`claim-circle-${netClaims}`} value={netClaims} />
          }
        >
          {hideText ? (
            <ClaimCircle key={`claim-circle-${netClaims}`} value={netClaims} />
          ) : (
            'VIEW & MINT'
          )}
        </Button>
      </Content>
    );
  }

  /* Connected to GitHub, but NO open claims */
  return (
    <Content className={className}>
      <ConnectedButton
        onClick={() => setIsOpen(true)}
        variant="outline"
        leftIcon={!hideText && <GoMarkGithub size={16} />}
        hideText={hideText}
      >
        {hideText ? <GoMarkGithub size={16} /> : 'NONE TO MINT'}
      </ConnectedButton>
    </Content>
  );
};
