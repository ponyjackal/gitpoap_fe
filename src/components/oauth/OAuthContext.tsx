import React, { createContext, useContext } from 'react';
import { useGithubAuth } from './hooks/useGithubAuth';
import { useDiscordAuth } from './hooks/useDiscordAuth';

type OAuthContextData = {
  github: {
    disconnect: () => void;
    authorize: () => void;
  };
  discord: {
    disconnect: () => void;
    authorize: () => void;
  };
};

const OAuthContext = createContext<OAuthContextData>({} as OAuthContextData);

export const useOAuthContext = () => {
  return useContext(OAuthContext);
};

type Props = {
  children: React.ReactNode;
};

export const OAuthProvider = ({ children }: Props) => {
  const github = useGithubAuth();
  const discord = useDiscordAuth();

  return (
    <OAuthContext.Provider
      value={{
        github: { ...github },
        discord: { ...discord },
      }}
    >
      {children}
    </OAuthContext.Provider>
  );
};
