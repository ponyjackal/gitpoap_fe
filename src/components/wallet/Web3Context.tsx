import { createContext, useCallback, useContext, useMemo, useState, useEffect } from 'react';
import { useWeb3React } from '@web3-react/core';
import { useDisclosure, useLocalStorage } from '@mantine/hooks';
import { JsonRpcSigner } from '@ethersproject/providers';
import detectEthereumProvider from '@metamask/detect-provider';
import { DateTime } from 'luxon';
import { useRefreshTokens } from '../../hooks/useRefreshTokens';
import { useTokens } from '../../hooks/useTokens';
import { useApi } from '../../hooks/useApi';
import { useIndexedDB, IndexDBStatus } from '../../hooks/useIndexedDB';
import { SignatureType } from '../../types';
import { AuthenticateResponse } from '../../lib/api/auth';
import { sign, generateSignatureData } from '../../lib/api/utils';
import { ONE_MONTH_IN_S, FIVE_MINUTES_IN_S } from '../../constants';
import { connectors } from '../../connectors';

type Props = {
  children: React.ReactNode;
};

export enum ConnectionStatus {
  CONNECTED_TO_WALLET = 'connected-to-wallet' /* Connected to wallet & authenticated */,
  CONNECTING_WALLET = 'connecting-wallet' /* Connecting to wallet & authenticating*/,
  DISCONNECTING = 'disconnecting' /* Disconnecting from wallet */,
  DISCONNECTED = 'disconnected' /* Not connected to any wallet */,
  UNINITIALIZED = 'uninitialized' /* Wallet connection hasn't been attempted yet */,
}

type onChainProvider = {
  address: string | null;
  setAddress: (address: string) => void;
  ensName: string | null;
  connectionStatus: ConnectionStatus;
  setConnectionStatus: (connectionStatus: ConnectionStatus) => void;
  disconnectWallet: () => void;
  isModalOpened: boolean;
  closeModal: () => void;
  handleConnect: () => void;
  isMetaMaskInstalled: boolean;
};

type Web3ContextState = {
  onChainProvider: onChainProvider;
} | null;

const Web3Context = createContext<Web3ContextState>(null);

export const useWeb3Context = () => {
  const web3Context = useContext(Web3Context);

  if (!web3Context) {
    throw new Error(
      'useWeb3Context() can only be used inside of <Web3ContextProvider />, ' +
        'please declare it at a higher level.',
    );
  }

  const { onChainProvider } = web3Context;

  return useMemo<onChainProvider>(() => {
    return { ...onChainProvider };
  }, [onChainProvider]);
};

export const Web3ContextProvider = (props: Props) => {
  const [connectionStatus, setConnectionStatus] = useState<ConnectionStatus>(
    ConnectionStatus.UNINITIALIZED,
  );
  const [address, setAddress] = useState<string | null>(null);

  const [isMetaMaskInstalled, setIsMetaMaskInstalled] = useState<boolean>(false);

  const { deactivate, account, library, activate } = useWeb3React();
  const isConnected = typeof account === 'string' && !!library;

  const [isModalOpened, { close: closeModal, open: openModal }] = useDisclosure(false);

  const api = useApi();

  const { setAccessToken, setRefreshToken, tokens, payload, refreshTokenPayload } = useTokens();

  const [provider] = useLocalStorage<string | null>({
    key: 'provider',
    defaultValue: null,
  });

  const {
    value: signature,
    setValue: setSignature,
    status: signatureStatus,
  } = useIndexedDB(account ?? '', null);

  /* This hook can only be used once here ~ it contains token refresh logic */
  useRefreshTokens();

  const disconnectWallet = useCallback(() => {
    deactivate();

    setConnectionStatus(ConnectionStatus.DISCONNECTED);
    setAddress('');
    setRefreshToken(null);
    setAccessToken(null);
  }, [deactivate, setConnectionStatus, setRefreshToken, setAccessToken, setAddress]);

  const authenticate = useCallback(
    async (signature: SignatureType) => {
      const authData: AuthenticateResponse | null = await api.auth.authenticate(signature);

      if (!authData) {
        // update connection status
        setConnectionStatus(ConnectionStatus.UNINITIALIZED);
        // update signature
        setSignature(null);
        return;
      }

      // set signature data into IndexedDB
      setSignature({
        ...authData.signatureData,
      });
      // update connection status
      setConnectionStatus(ConnectionStatus.CONNECTED_TO_WALLET);
      setAddress(authData.signatureData.address);
      // update tokens
      setAccessToken(authData.tokens.accessToken);
      setRefreshToken(authData.tokens.refreshToken);
    },
    [setConnectionStatus, setSignature, setAddress, api.auth, setAccessToken, setRefreshToken],
  );

  const authenticateWithoutSignature = useCallback(async () => {
    const signer: JsonRpcSigner = library.getSigner();
    const address = await signer.getAddress();
    const signatureData = generateSignatureData(address);
    const signatureString = await sign(signer, signatureData.message);

    if (!signatureString) return;

    await authenticate({
      ...signatureData,
      signature: signatureString,
      address,
    });
  }, [library, authenticate]);

  const authenticateWithSignature = useCallback(
    async (signature: SignatureType) => {
      await authenticate(signature);
    },
    [authenticate],
  );

  // check if metamask is installed
  useEffect(() => {
    if (typeof window === 'undefined') {
      return;
    }

    async function checkForMetaMask() {
      const provider = await detectEthereumProvider({
        timeout: 1000,
        mustBeMetaMask: true,
      });

      if (provider) {
        setIsMetaMaskInstalled(true);
      } else {
        setIsMetaMaskInstalled(false);
      }
    }

    void checkForMetaMask();
  }, []);

  // handle auth when account has changed
  useEffect(() => {
    // if wallet is not connected, do nothing
    if (!isConnected) return;
    // do nothing if signature is not loaded yet from indexedDB
    if (signatureStatus !== IndexDBStatus.LOADED || (signature && signature.address !== account))
      return;
    // if wallet is connecting, do nothing
    if (connectionStatus === ConnectionStatus.CONNECTING_WALLET) return;
    // if wallet is connected signed by current address, do nothing
    if (
      connectionStatus === ConnectionStatus.CONNECTED_TO_WALLET &&
      address?.toLowerCase() === account.toLowerCase()
    )
      return;

    // now that we go through authentication
    // set connection status as connecting
    setConnectionStatus(ConnectionStatus.CONNECTING_WALLET);

    if (signature) {
      void authenticateWithSignature(signature);
    } else {
      // we ask to sign a signature only if there is no signature for connected address
      void authenticateWithoutSignature();
    }
  }, [
    account,
    address,
    isConnected,
    signature,
    connectionStatus,
    signatureStatus,
    authenticateWithSignature,
    authenticateWithoutSignature,
    setConnectionStatus,
  ]);

  useEffect(() => {
    if (tokens?.accessToken === null && tokens?.refreshToken === null) {
      disconnectWallet();
    }
  }, [tokens, disconnectWallet]);

  // handle connect account if we have valid token in localstorage
  useEffect(() => {
    const connectToCachedProvider = async () => {
      // check if token is still not expired, connection status is uninitialized
      if (tokens?.accessToken && payload && connectionStatus === ConnectionStatus.UNINITIALIZED) {
        const accessTokenExp = payload?.exp;
        if (accessTokenExp) {
          const isExpired = DateTime.now().toUnixInteger() + FIVE_MINUTES_IN_S > accessTokenExp;
          if (!isExpired) {
            // get cached connector
            let cachedConnector = null;
            if (provider === 'injected') {
              cachedConnector = connectors.injected;
            } else if (provider === 'coinbase') {
              cachedConnector = connectors.coinbaseWallet;
            }

            if (!cachedConnector) return;

            setConnectionStatus(ConnectionStatus.CONNECTING_WALLET);

            void activate(cachedConnector);
            setAddress(payload.address);

            setConnectionStatus(ConnectionStatus.CONNECTED_TO_WALLET);
          }
        }
      }
    };

    void connectToCachedProvider();
  }, [tokens, payload, connectionStatus, activate, account, provider]);

  const refreshToken = useCallback(
    async (address: string) => {
      const tokens = await api.auth.refresh();

      if (tokens?.accessToken && tokens?.refreshToken) {
        setAccessToken(tokens.accessToken);
        setRefreshToken(tokens.refreshToken);

        // update connection status
        setConnectionStatus(ConnectionStatus.CONNECTED_TO_WALLET);
        setAddress(address);
      } else {
        disconnectWallet();
      }
    },
    [api.auth, setAccessToken, setRefreshToken, setConnectionStatus, setAddress, disconnectWallet],
  );

  const handleConnect = useCallback(async () => {
    // if previous access token exists, we check if refresh token is valid or not
    // if valid, we use it to refresh access token, no need to ask to sign
    const issuedAt = refreshTokenPayload?.iat ?? 0;
    const isExpired = DateTime.now().toUnixInteger() >= issuedAt + ONE_MONTH_IN_S;
    if (tokens && payload?.address && !isExpired && tokens.refreshToken) {
      // use existing refresh token to refresh access token
      setConnectionStatus(ConnectionStatus.CONNECTING_WALLET);
      void refreshToken(payload.address);
    } else {
      // otherwise, open wallet connect modal
      openModal();
    }
  }, [tokens, payload, openModal, refreshToken, setConnectionStatus, refreshTokenPayload?.iat]);

  const onChainProvider = useMemo(
    () => ({
      address,
      setAddress,
      connectionStatus,
      setConnectionStatus,
      ensName: payload?.ensName ?? null,
      disconnectWallet,
      handleConnect,
      isModalOpened,
      closeModal,
      isMetaMaskInstalled,
    }),
    [
      address,
      setAddress,
      connectionStatus,
      setConnectionStatus,
      payload?.ensName,
      disconnectWallet,
      handleConnect,
      isModalOpened,
      closeModal,
      isMetaMaskInstalled,
    ],
  );

  return <Web3Context.Provider value={{ onChainProvider }}>{props.children}</Web3Context.Provider>;
};
