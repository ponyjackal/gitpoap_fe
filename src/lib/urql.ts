import { initUrqlClient } from 'next-urql';
import {
  dedupExchange,
  cacheExchange,
  fetchExchange,
  ssrExchange,
  ClientOptions,
  createClient,
} from 'urql';
import { GITPOAP_API_URL } from '../environment';
import { getAccessToken } from '../hooks/useTokens';

export const urqlClientOptions: ClientOptions = {
  url: `${GITPOAP_API_URL}/graphql`,
  fetchOptions: () => {
    const accessToken = getAccessToken();
    return {
      headers: {
        Authorization: `Bearer ${accessToken ?? null}`,
      },
    };
  },
};

export const createSSRUrqlClient = () => {
  const ssrCache = ssrExchange({ isClient: false });
  const client = initUrqlClient(
    {
      ...urqlClientOptions,
      exchanges: [dedupExchange, cacheExchange, ssrCache, fetchExchange],
    },
    false,
  );

  return { client, ssrCache };
};

export const client = createClient(urqlClientOptions);
