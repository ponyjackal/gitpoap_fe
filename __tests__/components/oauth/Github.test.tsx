import { render } from '@testing-library/react';
import 'jest-styled-components';
import { RouterContext } from 'next/dist/shared/lib/router-context';
import { NextRouter } from 'next/router';
import { Provider as URQLProvider } from 'urql';
import { Web3ContextProvider } from '../../../src/components/wallet/Web3Context';
import { OAuthProvider } from '../../../src/components/oauth/OAuthContext';
import { ConnectionButton } from '../../../src/components/oauth/ConnectionButton';

const mockClient = {
  executeQuery: jest.fn(() => {}),
  executeMutation: jest.fn(() => {}),
  executeSubscription: jest.fn(() => {}),
  /* eslint-disable @typescript-eslint/no-explicit-any */
} as any;

const mockRouter: NextRouter = {
  basePath: '',
  pathname: '/',
  route: '/',
  asPath: '/',
  query: {},
  push: () => {},
  replace: () => {},
  reload: () => {},
  back: () => {},
  prefetch: () => {},
  beforePopState: () => {},
  events: {
    on: () => {},
    off: () => {},
    emit: () => {},
  },
  isFallback: false,
  isLocaleDomain: false,
} as any;

describe('ConnectionButton Button', () => {
  it('renders a Button', () => {
    const { container } = render(
      <RouterContext.Provider value={mockRouter}>
        <Web3ContextProvider>
          <OAuthProvider>
            <URQLProvider value={mockClient}>
              <ConnectionButton />
            </URQLProvider>
          </OAuthProvider>
        </Web3ContextProvider>
      </RouterContext.Provider>,
    );
    const button = container.firstChild;

    expect(button).toHaveTextContent('Check Eligibility');

    expect(button).toBeInTheDocument();
    expect(button).toMatchSnapshot();
  });

  it('renders a button and hides the text when hideText is true', () => {
    const { container } = render(
      <RouterContext.Provider value={mockRouter}>
        <Web3ContextProvider>
          <OAuthProvider>
            <URQLProvider value={mockClient}>
              <ConnectionButton hideText />
            </URQLProvider>
          </OAuthProvider>
        </Web3ContextProvider>
      </RouterContext.Provider>,
    );
    const button = container.firstChild;

    /* Expect it to have no text */
    expect(button).toHaveTextContent('');
    expect(button).toBeInTheDocument();
    expect(button).toMatchSnapshot();
  });
});
