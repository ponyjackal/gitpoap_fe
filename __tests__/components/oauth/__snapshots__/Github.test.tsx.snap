// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ConnectionButton Button renders a Button 1`] = `
.c1 {
  -webkit-text-decoration: none;
  text-decoration: none;
  -webkit-transition: 200ms fill ease,200ms stroke ease,200ms color ease;
  transition: 200ms fill ease,200ms stroke ease,200ms color ease;
  height: 1.125rem;
  width: 1.125rem;
  color: #838495;
}

.c1 path {
  -webkit-transition: 200ms fill ease;
  transition: 200ms fill ease;
}

.c0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: row;
  -ms-flex-direction: row;
  flex-direction: row;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  color: white;
}

.c2 path {
  fill: white;
}

.emotion-2 {
  -webkit-tap-highlight-color: transparent;
  font-family: -apple-system,BlinkMacSystemFont,Segoe UI,Roboto,Helvetica,Arial,sans-serif,Apple Color Emoji,Segoe UI Emoji;
  cursor: pointer;
  border: 0;
  padding: 0;
  -webkit-appearance: none;
  -moz-appearance: none;
  -ms-appearance: none;
  appearance: none;
  font-size: 16px;
  background-color: transparent;
  text-align: left;
  color: #000;
  -webkit-text-decoration: none;
  text-decoration: none;
  box-sizing: border-box;
  height: 36px;
  padding-left: 12px;
  padding-right: 18px;
  font-family: -apple-system,BlinkMacSystemFont,Segoe UI,Roboto,Helvetica,Arial,sans-serif,Apple Color Emoji,Segoe UI Emoji;
  -webkit-tap-highlight-color: transparent;
  display: inline-block;
  width: auto;
  border-radius: 4px;
  font-weight: 600;
  position: relative;
  line-height: 1;
  font-size: 14px;
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
  cursor: pointer;
  border: 1px solid transparent;
  background-color: #228be6;
  color: #fff;
}

.emotion-2:focus {
  outline-offset: 2px;
  outline: 2px solid #339af0;
}

.emotion-2:focus:not(:focus-visible) {
  outline: none;
}

.emotion-2:focus {
  outline-offset: 2px;
  outline: 2px solid #339af0;
}

.emotion-2:focus:not(:focus-visible) {
  outline: none;
}

@media (hover: hover) {
  .emotion-2:hover {
    background-color: #1c7ed6;
  }
}

@media (hover: none) {
  .emotion-2:active {
    background-color: #1c7ed6;
  }
}

.emotion-2:active {
  -webkit-transform: translateY(1px);
  -moz-transform: translateY(1px);
  -ms-transform: translateY(1px);
  transform: translateY(1px);
}

.emotion-2:disabled,
.emotion-2[data-disabled] {
  border-color: transparent;
  background-color: #e9ecef;
  color: #adb5bd;
  cursor: not-allowed;
  background-image: none;
  pointer-events: none;
}

.emotion-2:disabled:active,
.emotion-2[data-disabled]:active {
  -webkit-transform: none;
  -moz-transform: none;
  -ms-transform: none;
  transform: none;
}

.emotion-2[data-loading] {
  pointer-events: none;
}

.emotion-2[data-loading]::before {
  content: "";
  position: absolute;
  top: -1px;
  left: -1px;
  right: -1px;
  bottom: -1px;
  background-color: rgba(255, 255, 255, .5);
  border-radius: 4px;
  cursor: not-allowed;
}

.emotion-3 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -ms-flex-pack: center;
  -webkit-justify-content: center;
  justify-content: center;
  height: 100%;
  overflow: visible;
}

.emotion-7 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  margin-right: 10px;
}

.emotion-8 {
  white-space: nowrap;
  height: 100%;
  overflow: hidden;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
}

<div
  class="c0"
>
  <button
    class="emotion-0 emotion-1 emotion-2"
    data-button="true"
    type="button"
  >
    <div
      class="emotion-3 emotion-4"
    >
      <span
        class="emotion-5 emotion-6 emotion-7"
      >
        <svg
          class="c1 c2"
          fill="none"
          height="17"
          viewBox="0 0 15 17"
          width="15"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            clip-rule="evenodd"
            d="M12.383 11.144l1.691 3.092c.176.291 0 .496-.408.496l-1.488.058c-.32 0-.73.292-.846.554l-.788 1.46a.314.314 0 01-.583 0s-1.05-1.926-1.692-3.18c.166-.048.293-.115.397-.17l.07-.035 3.646-2.275zM.539 14.236l1.692-3.092 3.617 2.275c.054.022.104.047.156.074.089.046.182.094.31.13a173.077 173.077 0 01-1.691 3.18.315.315 0 01-.584 0l-.787-1.459c-.117-.291-.525-.554-.847-.554l-1.487-.058c-.438 0-.496-.234-.38-.496zM13.345 8.43c0 .817-.496 1.663-1.225 2.072l-3.588 1.984a2.81 2.81 0 01-1.226.292c-.408 0-.846-.059-1.225-.292l-3.588-2.013c-.73-.408-1.225-1.254-1.225-2.071V4.376c0-.817.496-1.663 1.225-2.071L6.081.29C6.43.116 6.869 0 7.306 0c.409 0 .788.058 1.226.291l3.588 2.013c.787.41 1.225 1.196 1.225 2.071V8.43zm-6.943 2.393c.291.175.583.204.904.204.32 0 .641-.059.904-.204l2.597-1.459c.554-.32.904-.904.904-1.517V4.93c0-.671-.321-1.225-.904-1.517L8.21 1.954c-.32-.175-.583-.204-.904-.204-.321 0-.642.088-.904.204L3.805 3.413c-.554.32-.904.904-.904 1.517v2.917c0 .613.379 1.226.904 1.517l2.597 1.459zm-.513-5.694a.365.365 0 00-.626-.249L4.214 5.96a.365.365 0 00-.003.504l1.05 1.109a.365.365 0 00.528-.502l-.808-.854.805-.829a.365.365 0 00.103-.259zm3.052-.331a.364.364 0 00-.118.594l.825.825-.828.85a.365.365 0 10.522.51l1.079-1.109a.365.365 0 00-.003-.512l-1.08-1.08a.364.364 0 00-.397-.078zm-.77.047a.365.365 0 00-.12-.078h.001a.364.364 0 00-.475.201l-.933 2.305a.364.364 0 10.676.274l.933-2.305a.365.365 0 00-.082-.397z"
            fill="#838495"
            fill-rule="evenodd"
          />
        </svg>
      </span>
      <span
        class="emotion-8 emotion-9"
      >
        Check Eligibility
      </span>
    </div>
  </button>
</div>
`;

exports[`ConnectionButton Button renders a button and hides the text when hideText is true 1`] = `
.c1 {
  -webkit-text-decoration: none;
  text-decoration: none;
  -webkit-transition: 200ms fill ease,200ms stroke ease,200ms color ease;
  transition: 200ms fill ease,200ms stroke ease,200ms color ease;
  height: 1.125rem;
  width: 1.125rem;
  color: #838495;
}

.c1 path {
  -webkit-transition: 200ms fill ease;
  transition: 200ms fill ease;
}

.c0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: row;
  -ms-flex-direction: row;
  flex-direction: row;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  color: white;
}

.c2 path {
  fill: white;
}

.emotion-2 {
  -webkit-tap-highlight-color: transparent;
  font-family: -apple-system,BlinkMacSystemFont,Segoe UI,Roboto,Helvetica,Arial,sans-serif,Apple Color Emoji,Segoe UI Emoji;
  cursor: pointer;
  border: 0;
  padding: 0;
  -webkit-appearance: none;
  -moz-appearance: none;
  -ms-appearance: none;
  appearance: none;
  font-size: 16px;
  background-color: transparent;
  text-align: left;
  color: #000;
  -webkit-text-decoration: none;
  text-decoration: none;
  box-sizing: border-box;
  height: 36px;
  padding-left: 18px;
  padding-right: 18px;
  font-family: -apple-system,BlinkMacSystemFont,Segoe UI,Roboto,Helvetica,Arial,sans-serif,Apple Color Emoji,Segoe UI Emoji;
  -webkit-tap-highlight-color: transparent;
  display: inline-block;
  width: auto;
  border-radius: 4px;
  font-weight: 600;
  position: relative;
  line-height: 1;
  font-size: 14px;
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
  cursor: pointer;
  border: 1px solid transparent;
  background-color: #228be6;
  color: #fff;
}

.emotion-2:focus {
  outline-offset: 2px;
  outline: 2px solid #339af0;
}

.emotion-2:focus:not(:focus-visible) {
  outline: none;
}

.emotion-2:focus {
  outline-offset: 2px;
  outline: 2px solid #339af0;
}

.emotion-2:focus:not(:focus-visible) {
  outline: none;
}

@media (hover: hover) {
  .emotion-2:hover {
    background-color: #1c7ed6;
  }
}

@media (hover: none) {
  .emotion-2:active {
    background-color: #1c7ed6;
  }
}

.emotion-2:active {
  -webkit-transform: translateY(1px);
  -moz-transform: translateY(1px);
  -ms-transform: translateY(1px);
  transform: translateY(1px);
}

.emotion-2:disabled,
.emotion-2[data-disabled] {
  border-color: transparent;
  background-color: #e9ecef;
  color: #adb5bd;
  cursor: not-allowed;
  background-image: none;
  pointer-events: none;
}

.emotion-2:disabled:active,
.emotion-2[data-disabled]:active {
  -webkit-transform: none;
  -moz-transform: none;
  -ms-transform: none;
  transform: none;
}

.emotion-2[data-loading] {
  pointer-events: none;
}

.emotion-2[data-loading]::before {
  content: "";
  position: absolute;
  top: -1px;
  left: -1px;
  right: -1px;
  bottom: -1px;
  background-color: rgba(255, 255, 255, .5);
  border-radius: 4px;
  cursor: not-allowed;
}

.emotion-3 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -ms-flex-pack: center;
  -webkit-justify-content: center;
  justify-content: center;
  height: 100%;
  overflow: visible;
}

.emotion-5 {
  white-space: nowrap;
  height: 100%;
  overflow: hidden;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
}

<div
  class="c0"
>
  <button
    class="emotion-0 emotion-1 emotion-2"
    data-button="true"
    type="button"
  >
    <div
      class="emotion-3 emotion-4"
    >
      <span
        class="emotion-5 emotion-6"
      >
        <svg
          class="c1 c2"
          fill="none"
          height="17"
          viewBox="0 0 15 17"
          width="15"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            clip-rule="evenodd"
            d="M12.383 11.144l1.691 3.092c.176.291 0 .496-.408.496l-1.488.058c-.32 0-.73.292-.846.554l-.788 1.46a.314.314 0 01-.583 0s-1.05-1.926-1.692-3.18c.166-.048.293-.115.397-.17l.07-.035 3.646-2.275zM.539 14.236l1.692-3.092 3.617 2.275c.054.022.104.047.156.074.089.046.182.094.31.13a173.077 173.077 0 01-1.691 3.18.315.315 0 01-.584 0l-.787-1.459c-.117-.291-.525-.554-.847-.554l-1.487-.058c-.438 0-.496-.234-.38-.496zM13.345 8.43c0 .817-.496 1.663-1.225 2.072l-3.588 1.984a2.81 2.81 0 01-1.226.292c-.408 0-.846-.059-1.225-.292l-3.588-2.013c-.73-.408-1.225-1.254-1.225-2.071V4.376c0-.817.496-1.663 1.225-2.071L6.081.29C6.43.116 6.869 0 7.306 0c.409 0 .788.058 1.226.291l3.588 2.013c.787.41 1.225 1.196 1.225 2.071V8.43zm-6.943 2.393c.291.175.583.204.904.204.32 0 .641-.059.904-.204l2.597-1.459c.554-.32.904-.904.904-1.517V4.93c0-.671-.321-1.225-.904-1.517L8.21 1.954c-.32-.175-.583-.204-.904-.204-.321 0-.642.088-.904.204L3.805 3.413c-.554.32-.904.904-.904 1.517v2.917c0 .613.379 1.226.904 1.517l2.597 1.459zm-.513-5.694a.365.365 0 00-.626-.249L4.214 5.96a.365.365 0 00-.003.504l1.05 1.109a.365.365 0 00.528-.502l-.808-.854.805-.829a.365.365 0 00.103-.259zm3.052-.331a.364.364 0 00-.118.594l.825.825-.828.85a.365.365 0 10.522.51l1.079-1.109a.365.365 0 00-.003-.512l-1.08-1.08a.364.364 0 00-.397-.078zm-.77.047a.365.365 0 00-.12-.078h.001a.364.364 0 00-.475.201l-.933 2.305a.364.364 0 10.676.274l.933-2.305a.365.365 0 00-.082-.397z"
            fill="#838495"
            fill-rule="evenodd"
          />
        </svg>
      </span>
    </div>
  </button>
</div>
`;
